projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,201 @@
+#include "ads1015.h"
+#include <status.h>
+#include <stdio.h>
+#include "ads1015_def.h"
+#include "gpio_it.h"
+
+static void prv_check_status(StatusCode code) {
+  if (!code) {
+    status_msg(code, "failed");
+  }
+}
+
+// reset the internal registers and enter a power-down state
+static StatusCode prv_reset(ADS1015Data *data) {
+  uint8_t reset = ADS1015_RESET_BYTE;
+  StatusCode code = i2c_write(data->i2c_port, ADS1015_I2C_GENERAL_CALL, &reset, sizeof(reset));
+  prv_check_status(code);
+  status_ok_or_return(code);
+  return STATUS_CODE_OK;
+}
+
+// writes to register given upper and lower bytes
+static StatusCode prv_setup_register(ADS1015Data *data, uint8_t reg, uint8_t msb, uint8_t lsb) {
+  uint8_t ads1015_setup_register[] = { reg, msb, lsb };
+  StatusCode code = i2c_write(data->i2c_port, data->i2c_addr | ADS1015_I2C_ADDRESS_GND,
+                              ads1015_setup_register, SIZEOF_ARRAY(ads1015_setup_register));
+  prv_check_status(code);
+  status_ok_or_return(code);
+  return STATUS_CODE_OK;
+}
+
+// reads the register and stores the value in the given array(2 bytes)
+static StatusCode prv_read(I2CPort i2c_port, ADS1015Address i2c_addr, uint8_t reg, uint8_t *read,
+                           size_t size) {
+  StatusCode code = i2c_write(i2c_port, i2c_addr | ADS1015_I2C_ADDRESS_GND, &reg, sizeof(reg));
+  prv_check_status(code);
+  status_ok_or_return(code);
+  code = i2c_read(i2c_port, i2c_addr | ADS1015_I2C_ADDRESS_GND, read, size);
+  prv_check_status(code);
+  status_ok_or_return(code);
+  return STATUS_CODE_OK;
+}
+
+// gets current channel based on the upper byte of config register
+static ADS1015Channel prv_get_channel(uint8_t config_register_msb) {
+  uint8_t channel = (config_register_msb >> 4) & 0x7;  // 4 <= channel <= 7
+  channel -= 4;  // adjusts the register value to channels' enum: 0 <= channel <= 3
+  return channel;
+}
+
+// switches to the the given channel by writing to config reg
+static StatusCode prv_set_channel(ADS1015Data *data, ADS1015Channel channel) {
+  if (channel >= NUM_ADS1015_CHANNELS) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+  uint8_t config_register_msb_channels[] = { CONFIG_REGISTER_MSB_0, CONFIG_REGISTER_MSB_1,
```

Comment:
Ah, I see what you're doing with the MSBs. Instead of doing this, why not just OR the channel with the data rate, single shot, etc? You could make the common settings a define if you wanted to. I think that would be much easier to understand and remove the need for this lookup table.